tools/server/server-context.cpp

```diff
@@ -1124,6 +1161,79 @@ struct server_context_impl {
             SLT_DBG(slot, "stopped by limit, n_decoded = %d, n_predict = %d\n", slot.n_decoded, slot.task->params.n_predict);
         }
 
+        const int32_t reasoning_budget = (slot.task && slot.task->params.reasoning_budget_override.has_value())
+            ? slot.task->params.reasoning_budget_override.value()
+            : params_base.reasoning_budget;
+
+        // check reasoning budget limit
+        // Track reasoning tokens when we're inside thinking blocks (<think>...</think> or similar)
+        // When the budget is exceeded we enqueue the closing tag tokens so they get sent to the client
+        // and fed back into the model before continuing normal generation
+        if (slot.has_next_token && reasoning_budget > 0) {
+            // Check if we've entered or exited reasoning mode
+            for (const auto & [start_tag, end_tag] : kReasoningThinkMarkers) {
+                size_t start_pos = slot.generated_text.rfind(start_tag);
+                size_t end_pos   = slot.generated_text.rfind(end_tag);
```

Comment:
IMO this can decrease the performance quite a lot, especially on long `generated_text`. Wondering if we should only support this feature for models with reasoning tag as one single token.

Reply:
Indeed.
In my opinion, we should embed somewhere in the GGUF informations about the thinking tags.
This would remove the need to do a dirty hardcoding like I did.
Also, the rfind could be limited to the closing token's length.

Reply:
Then I think allowing GGUF to hold metadata about that would be a good thing. It could even be reported.